Support next by distance and label as number

JezPrime-trp opened the pull request:
This pull request adds the following capabilities to the qoomon plugin

1. Ability to extract a number from the label (if the label is numeric) – note that this needs to be used with care as a non-numeric label fed into this will cause a build failure – this is because we wish to use (on master) versions of the form 1.2.3-456 (where 456 is auto-generated based on previous build number – more on this shortly)
2. As well as “next” for each property, a “next by distance” where the value is incremented by the distance since the last matching tag. If there is no matching tag then the total number of commits on the repository is used instead.
 
I have a few notes/queries relating to this. We originally did this against an older version of this plugin and I have some tests (which mainly exercised the describe functionality which we added as you have added it more recently) but which covered usage of the placeholders. I notice that you don’t explicitly test usage of placeholders, but I will be very happy to add tests specifically around these placeholders if desired – I was trying to stay close to the existing conventions. Please just let me know if this is preferred.

Secondly, I notice that you have “version.label” but “describe.version.version.label” – is this intentional? My submission currently assumes the inconsistency was unintentional and I’ve changed to “describe.version.label” but again will be happy to revert this piece and make the new properties consistent with “describe.version.version.label” – again please just let me know.

Finally, outside the scope of the functionality, I updated the README to refer to the current version of the plugin - I hope this is ok.

qoomon commented:
Thank you very much for your PR. I'll review and merge it within the next week.

qoomon commented:
Maybe `.asNumber` instead of `.number` it may be easier to understand that way. WDYT?

qoomon commented on README.md:
```diff
@@ -185,6 +185,7 @@ e.g `${dirty:-SNAPSHOT}` resolves to `-SNAPSHOT` instead of `-DIRTY`
     - `${version.patch.next}` the `${version.patch}` increased by 1 e.g. '4'
   - `${version.label}` the version label of `${version}` e.g. 'SNAPSHOT'
     - `${version.label.prefixed}` like `${version.label}` with label separator e.g. '-SNAPSHOT'
+    - `${version.label.number}` used to extract the label as a number (build will fail if non-numeric label encountered for this)
```
I think we don't need the as number placeholder for the project version label. If you need it please add documentation for `version.label.number.next` and `version.label.number.nextByDistance`

qoomon replied:
following description should be enough
- ${version.label.number} the ${version.label} as number

JezPrime-trp replied:
Happy to remove this, will let you know when I've pushed its removal.

JezPrime-trp commented on README.md:
```diff
@@ -248,10 +249,17 @@ e.g `${dirty:-SNAPSHOT}` resolves to `-SNAPSHOT` instead of `-DIRTY`
       - `${describe.tag.version.patch.next}` the `${describe.tag.version.patch}` increased by 1 e.g. '4'
+      - `${describe.tag.version.patch.nextByDistance}` the `${describe.tag.version.patch}` increased by `${describe.distance}` + 1 e.g. '2'
+    - `${describe.tag.version.label}` the label version component of `${describe.tag.version}` e.g. 'SNAPSHOT'
+    - `${describe.tag.version.label.number}` the `${describe.tag.version.label}` assuming numeric e,g, '5', will cause build failure if a non-numeric label is found
```
Will update and let you know when I've done so.

JezPrime-trp replied:
I've pushed changes which I think accommodate all the comments you made, please let me know if there's more to do.

qoomon commented:
Or we could get rid of `.number` and just offer `...label.next` and `...label.nextPlusDistance`

JezPrime-trp commented:
> Maybe `.asNumber` or `.asInt` instead of `.number` it may be easier to understand that way. WDYT?

I think this makes sense and offers clarity. I'm also happy to remove the .number altogether - it does give the "next" for label some meaning but I think I can deal with that by updating the doc.

Should it be describe.version.version.label for everything or describe.version.version.label...? Prior to my commit it's describe.version.version.label but just version.label. My change as is switches to describe.version.version.label but am I risking causing a break for people already using it?

JezPrime-trp commented:
> Or we could get rid of `.number` and just offer `...label.next` and `...label.nextPlusDistance`

I think I'm happy to remove '.number' - we can always re-add it if someone comes up with a use-case - it's not needed for my usage and was mainly in there to show how we get to that (but I will update the README to cover that).

JezPrime-trp commented:
Thank you very much for your quick and constructive review

JezPrime-trp commented:
I've updated and think I've accommodated all discussions so far in the changes - thank you very much for the quick review. I've also added the changes to https://github.com/qoomon/maven-git-versioning-extension/pull/209

qoomon commented:
Would be nice if you can change following things, then we're good to go.
* rename `nextByDistance` to `nextPlusDistance` 
* and remove `.asInt` (former `.number`) and change it to `...label.next` style

qoomon commented:
LGTM. However one last question, what is the use case for `....nextPlusDistance` why not just `describe.distance`?

JezPrime-trp commented:
> LGTM. However one last question, what is the use case for `....nextPlusDistance` why not just `describe.distance`?

In our case we want to keep that "build number" unique, not just unique within a given major/minor/patch combination - so no two commits should ever share the same build number. In practice it is possible not to use (label + distance) in this case as properly used it should always be "number of commits + 1" but grabbing it from the previous tag reduces the amount of commit-counting required in deep repositories. If no tag is used then it resolves to "describe.distance" effectively. I appreciate that not everyone wants a use-case that matches this, but following our team discussions we are very keen to have unique build numbers. I hope this is clear and makes sense.

qoomon commented:
Oh I think I am wrong my example would not be uniq if your distance on two branches are the same. You would have this problem as well with your solution. What you might want ost something like `...-${env.BUILD_NUMBER:-local}` (adjust BUILD_NUMBER to your CI environment)

JezPrime-trp commented:
> if you just use a pattern like `${describe.tag.version.core}-${describe.distance}` it would be uniq. What do I missing?

We want the label component (which represents build number) to be unique, not just the whole version. So our default stamp on master would be 
${describe.tag.version.core}-${describe.tag.version.nextPlusDistance} - effectively we want an "all time" build number not a "within this release' build number.

qoomon commented:
If you use branches there is no way in creating a unique incrementing number. a commit hash is the only unique identifier.

JezPrime-trp commented:
> Oh I think I am wrong my example would not be uniq if your distance on two branches are the same. You would have this problem as well with your solution. What you might want ost something like `...-${env.BUILD_NUMBER:-local}` (adjust BUILD_NUMBER to your CI environment)

I used this approach in a previous employment. Every so often for one service the jenkins config got reset (I don't know if the ops guys ever figured out why) and we then had to trigger lots of failed builds to get caught back up! So we're wary of this approach especially as we (our dev team) doesn't manage the build servers. We do however own our history.

JezPrime-trp commented:
> If you use branches there is no way in creating a unique incrementing number. a commit hash is the only unique identifier.

For branches we're less concerned as they are *not* release candidates. They'll use feature name (branch name) alongside the "proposed" build number in the resultant label so will be distinct from those on master. (We already do this)

qoomon commented:
One more question to totally understand your need for commit count on a branch. Why the total commit count on a branch is better than the relative commit count to a release tag?  (1.2.3-666 vs. 1.2.3-42)

Regarding LGTM I don't know either 😅

JezPrime-trp commented:
> One more question to totally understand your need for commit count on a branch. Why the total commit count on a branch is better than the relative commit count to a release tag? (1.2.3-666 vs. 1.2.3-42)

We only use the build number on the master branch - it's an internal convention we're targeting. I'm happy to go over it again with my team leader when I'm back in the office as it's his expressed preference. I actually added describe in an internal fork 2 years ago to support this function, so you have no idea how pleased I was that I could remove most of the code I added in the fork thanks to your changes since, but we have implemented using that approach (i.e. commit count on master is our build number, and we never duplicate build number between versions).

qoomon commented:
Hi @JezPrime-trp, i have some further questions to understand your issue:
* why is your next build number `lastBuildNumber + 1 + distance` ? Shouldn't it be enough to just add distance? 
* why not just `...label.next`?

JezPrime-trp commented:
> Hi @JezPrime-trp, i have some further questions to understand your issue:
> 
> * why is your next build number `lastBuildNumber + 1 + distance` ? Shouldn't it be enough to just add distance?
> * why not just `lastBuildNumber + 1` e.g. `...label.next`?

First question: I think you're right. We were trying to take account of uncommitted changes so "next" means the "next one after the build number derived from current commit". However, I don't think we care too much about accuracy in the "uncommitted" changes scenario (since by the time we really care about versions we're already in CI and therefore committed), and if we leave that code as is, in effect if we have a build tagged "1.0.1-12" and make one commit to it, then assuming we auto tag at this next commit we would get "1.0.1-14" not "1.0.1-13". I will amend this piece of the algorithm early next week. Note that at present we are not auto-tagging but this is in our plan, so thanks for raising this. 

On  your second question, I probably need to give you a bit more on our build/publish infrastructure, because I hope this will explain. Our build machines are really docker containers deployed and executed on demand, which means that the "build number" situation is actually worse than I outlined in the previous comment I made on the subject, based on a previous experience. However, this also means that two people can commit and have their changes merged to master in rapid succession potentially. Since the version is calculated during the gradle phase of the build, before we deploy to internal artifactory, there is a race condition if we only use +1. If we base the number on distance we do not experience this problem because although the builds will overlap, the commit distances since the last tag will be different. Because every build from master goes to artifactory, even if we haven't added a new tag (but even if we auto-tagged, that would probably only happen on build completion when we identify that we're ok to deploy), that race condition leads to uncertainty about which code we get in artifactory. Again in the scenario above, we would get two builds producing a build versioned 1.0.1-13 in artifactory, whereas basing on distance gives us 1.0.1-13 and 1.0.1-14. 

It's also worth keeping in mind that while our primary use-case is for master builds, as long as we don't overlap baseline versions on branches (e.g. if we need to make a 1.0 branch to support an older version , while pogressing 1.2 on master) if we base this from the latest 1.0 tag on master our build number tag will carry over and be seen on the branch, so within the version the build numbers will stay consistent, although they will no longer be unique between 1.0.1 branch and 1.0.2 on master). However just using distance would not achieve this it cannot guarantee unique build numbers even within a branch (e.g. you tag 1.0.1-5 and commit, you then go back to 1.0.1-1 which would already have gone to artifactory). I hope these examples make it clear why we're keen to use (label plus distance) rather than just (distance). I should have made it clear to you earlier that we can get builds overlapping like this, and that every successful build leads to a publish to artifactory, so ambiguity is a problem for is that this approach overcomes.

qoomon commented:
Okay got your point. Makes sense to me and I'm willing to add these placeholders.

I would suggest to rename `.nextPlusDistance` to something like this `.plus.describe.distance` 

And it would be nice to offer `.plus.describe.distance` for `describe.tag.version.patch` placeholders as well, because I think a much more common use-case would be `v1.2.100` where `100` is the build number, because it follows semantic versioning. or you

JezPrime-trp commented:
> Okay got your point. Makes sense to me and I'm willing to add these placeholders.
> 
> I would suggest to rename `.nextPlusDistance` to something like this `.plus.describe.distance`
> 
> And it would be nice to offer `.plus.describe.distance` for `describe.tag.version.patch` placeholders as well, because I think a much more common use-case would be `v1.2.100` where `100` is the build number, because it follows semantic versioning. or you

I thought we were offering the equivalent. If not it's an oversight on my part (once we added the concept of a plus.distance, I added a version of it for each version component). I will amend the +1 and the naming and ensure that at the very least we're offering it for the patch component, if not all components. Hoping to push updates (including conflict resolution in README) today. Thanks again for taking the time to review and consider.

qoomon commented:
following placeholders would be nice I think. What do you think?
* `${describe.tag.version.label.plus.describe.distance}`
* `${describe.tag.version.label.next.plus.describe.distance}`
* `${describe.tag.version.patch.plus.describe.distance}`
* `${describe.tag.version.patch.next.plus.describe.distance}`

JezPrime-trp commented:
Those placeholders look good to me. Do you want any other version components treated the same way or is that the exhaustive list? 

I may add some integration-like tests around label plus describe distance in GitVersioningPluginTest to ensure I'm getting the results I expect, at least for the baseline describe.tag.version.label.plus.describe.distance - is that ok with you? Should also let you validate my expectations :)

qoomon commented:
The 4 placeholder mentioned are sufficient for me.

JezPrime-trp commented:
Thanks, we had provided the same functionality for the major and minor components for completeness. I'll remove those. I will be making a small change to the describe distance calculation so that it returns the number of commits when no matching tag is found, rather than number of commits - 1 (has to do with starting at -1). Will also make the same changes to the maven plugin. Hoping to have it all pushed this afternoon to both repos.

qoomon commented on src/main/java/me/qoomon/gitversioning/commons/GitUtil.java:
```diff
@@ -78,7 +78,9 @@ public static GitDescription describe(ObjectId revObjectId, Pattern tagPattern,
             }
 
-            return new GitDescription(revObjectId.getName(), "root", depth);
+            // If no matching tag found, then we should presumably return the number of commits on the branch
```
I think it would be simpler to start with `int depth = 0;` and move `depth++;`  below conditional return block in while loop(line 72) and just return `new GitDescription(revObjectId.getName(), "root", depth);` here.

JezPrime-trp replied:
I believe this is dealt with by my  latest commit - please re-review when you get time

qoomon commented on src/main/java/me/qoomon/gradle/gitversioning/GitVersioningPluginExtension.java:
```diff
@@ -554,6 +553,7 @@ private Map<String, Supplier<String>> generateGlobalFormatPlaceholderMap(GitSitu
             placeholderMap.put("describe.tag." + groupName, groupValue);
             placeholderMap.put("describe.tag." + groupName + ".slug", Lazy.by(() -> slugify(groupValue.get())));
         }
+        final Lazy<Integer> distance = Lazy.by(() -> description.get().getDistance());
```
please rename variable to `descriptionDistance`and move this statement next to where it is used (line 590)

JezPrime-trp replied:
Done - see latest commit

qoomon commented on src/main/java/me/qoomon/gradle/gitversioning/GitVersioningPluginExtension.java:
```diff
@@ -577,8 +579,13 @@ private Map<String, Supplier<String>> generateGlobalFormatPlaceholderMap(GitSitu
 
         placeholderMap.put("describe.tag.version.patch", Lazy.by(() -> notNullOrDefault(descriptionTagVersionComponents.get().group("patch"), "0")));
         placeholderMap.put("describe.tag.version.patch.next", Lazy.by(() -> increaseStringNumber(placeholderMap.get("describe.tag.version.patch").get())));
+        placeholderMap.put("describe.tag.version.patch.plus.describe.distance", Lazy.by(() -> increaseStringNumberBy(placeholderMap.get("describe.tag.version.patch").get(), distance.get() )));
```
move all `....describe.distance` placeholder below `placeholderMap.put("describe.distance",...` (line 590)

JezPrime-trp replied:
Done - see latest commit

qoomon commented on src/main/java/me/qoomon/gradle/gitversioning/GitVersioningPluginExtension.java:
```diff
@@ -577,8 +579,13 @@ private Map<String, Supplier<String>> generateGlobalFormatPlaceholderMap(GitSitu
+        placeholderMap.put("describe.tag.version.label.plus.describe.distance", Lazy.by(() -> increaseStringNumberBy(labelAsInteger.get(), distance.get())));
+        placeholderMap.put("describe.tag.version.label.next.plus.describe.distance", Lazy.by(() -> increaseStringNumberBy(labelAsInteger.get(), distance.get() + 1)));
 
         placeholderMap.put("describe.distance", Lazy.by(() -> String.valueOf(description.get().getDistance())));
```
make use of `descriptionDistanceand ` (`distance`) in here

JezPrime-trp replied:
Done - see latest commit

qoomon commented on src/main/java/me/qoomon/gradle/gitversioning/GitVersioningPluginExtension.java:
```diff
@@ -673,7 +686,11 @@ private static String slugify(String value) {
+        return increaseStringNumberBy(majorVersion, 1);
+    }
+
+    private static String increaseStringNumberBy(String toIncrease, long by) {
```
please rename to `increase(String number, long increment)`

JezPrime-trp replied:
Done - see latest commit

qoomon commented on src/main/java/me/qoomon/gradle/gitversioning/GitVersioningPluginExtension.java:
```diff
@@ -567,6 +567,8 @@ private Map<String, Supplier<String>> generateGlobalFormatPlaceholderMap(GitSitu
             return matcher;
         });
 
+        final Lazy<String> labelAsInteger = Lazy.by(() -> extractLabelNumber(descriptionTagVersionComponents));
```
I don't think we need this at all as well as we don't need `extractLabelNumber`. We could just use `placeholderMap.get("describe.tag.version.label").get()` and refactor `increaseStringNumberBy` to treat empty string as `0`

JezPrime-trp replied:
Done - see latest commit. Nice spot!

JezPrime-trp commented:
I've pushed a new commit which I think resolves your latest comments. Please re-review when you get time. I'm still waiting for our fork to be updated, but in the interim will keep responding to your comments as quickly as I can so that when our fork is up to date all I should need to do is get the conflict resolved. Thanks

qoomon commented:
@JezPrime-trp I resolved the conflict on my own. Thanks again for your work.

qoomon commented:
version `6.3.0` is released

JezPrime-trp commented:
Thank you very much for the review and merge of this functionality which will really help us (we were using a fork of an earlier version into which I'd added describe functionality).
